File: typescript/src/transport/delivery.ts
import type { DeliveryCounts, EventBuffer } from "./buffer.js";

export type DeliveryWorkerState = "local_only" | "idle" | "syncing" | "backoff" | "auth_failed" | "stopped";
export type DeliveryErrorOperation = "transport" | "authentication" | "conversion";

export interface DeliveryErrorEvent {
  occurredAt: Date;
  operation: DeliveryErrorOperation;
  errorType: string;
  message: string;
  retryable: boolean;
  consecutiveFailures: number;
}

export type DeliveryErrorCallback = (event: DeliveryErrorEvent) => void;

export interface DeliveryStatusOptions extends DeliveryCounts {
  enabled: boolean;
  workerState: DeliveryWorkerState;
  lastAttemptAt?: Date;
  lastSuccessAt?: Date;
  lastErrorAt?: Date;
  lastErrorType?: string;
  lastErrorMessage?: string;
  consecutiveFailures?: number;
  successfulBatches?: number;
  failedBatches?: number;
  deliveredRecords?: number;
  backoffSeconds?: number;
}

export class DeliveryStatus implements DeliveryStatusOptions {
  readonly enabled: boolean;
  readonly workerState: DeliveryWorkerState;
  readonly pendingEvents: number;
  readonly quarantinedEvents: number;
  readonly pendingTasks: number;
  readonly quarantinedTasks: number;
  readonly pendingOutcomes: number;
  readonly quarantinedOutcomes: number;
  readonly pendingRevenues: number;
  readonly quarantinedRevenues: number;
  readonly pendingProviderJobs: number;
  readonly quarantinedProviderJobs: number;
  readonly oldestPendingAt?: Date;
  readonly lastAttemptAt?: Date;
  readonly lastSuccessAt?: Date;
  readonly lastErrorAt?: Date;
  readonly lastErrorType?: string;
  readonly lastErrorMessage?: string;
  readonly consecutiveFailures: number;
  readonly successfulBatches: number;
  readonly failedBatches: number;
  readonly deliveredRecords: number;
  readonly backoffSeconds: number;

  constructor(options: DeliveryStatusOptions) {
    Object.assign(this, options);
    this.enabled = options.enabled; this.workerState = options.workerState;
    this.pendingEvents = options.pendingEvents; this.quarantinedEvents = options.quarantinedEvents;
    this.pendingTasks = options.pendingTasks; this.quarantinedTasks = options.quarantinedTasks;
    this.pendingOutcomes = options.pendingOutcomes;
    this.quarantinedOutcomes = options.quarantinedOutcomes; this.pendingRevenues = options.pendingRevenues;
    this.quarantinedRevenues = options.quarantinedRevenues;
    this.pendingProviderJobs = options.pendingProviderJobs;
    this.quarantinedProviderJobs = options.quarantinedProviderJobs;
    this.consecutiveFailures = options.consecutiveFailures ?? 0;
    this.successfulBatches = options.successfulBatches ?? 0;
    this.failedBatches = options.failedBatches ?? 0;
    this.deliveredRecords = options.deliveredRecords ?? 0;
    this.backoffSeconds = options.backoffSeconds ?? 0;
  }

  get pendingRecords(): number {
    return this.pendingEvents + this.pendingTasks + this.pendingOutcomes +
      this.pendingRevenues + this.pendingProviderJobs;
  }

  get quarantinedRecords(): number {
    return this.quarantinedEvents + this.quarantinedTasks + this.quarantinedOutcomes +
      this.quarantinedRevenues + this.quarantinedProviderJobs;
  }

  get healthy(): boolean {
    return this.workerState !== "auth_failed" && this.workerState !== "backoff" &&
      this.quarantinedRecords === 0;
  }
}

const callbacks = new Set<DeliveryErrorCallback>();

export function onDeliveryError(callback: DeliveryErrorCallback): DeliveryErrorCallback {
  if (typeof callback !== "function") throw new TypeError("delivery error callback must be callable");
  callbacks.add(callback);
  return callback;
}

export function removeDeliveryErrorCallback(callback: DeliveryErrorCallback): boolean {
  return callbacks.delete(callback);
}

export function emitDeliveryError(event: DeliveryErrorEvent): void {
  for (const callback of [...callbacks]) {
    try { callback(event); } catch { /* observability must not break delivery */ }
  }
}

const emptyCounts = (): DeliveryCounts => ({
  pendingEvents: 0, quarantinedEvents: 0, pendingTasks: 0, quarantinedTasks: 0,
  pendingOutcomes: 0, quarantinedOutcomes: 0, pendingRevenues: 0,
  quarantinedRevenues: 0, pendingProviderJobs: 0, quarantinedProviderJobs: 0,
});

export function localDeliveryStatus(buffer?: EventBuffer): DeliveryStatus {
  return new DeliveryStatus({
    ...(buffer?.deliveryCounts() ?? emptyCounts()), enabled: false, workerState: "local_only",
  });
}

File: python/src/dexcost/delivery.py
"""Observable delivery health for DexCost's durable store-and-forward queue."""

from __future__ import annotations

import logging
import threading
from collections.abc import Callable
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Literal

_LOG = logging.getLogger(__name__)

DeliveryWorkerState = Literal[
    "local_only",
    "idle",
    "syncing",
    "backoff",
    "auth_failed",
    "stopped",
]
DeliveryErrorOperation = Literal["transport", "authentication", "conversion"]


@dataclass(frozen=True)
class DeliveryErrorEvent:
    """One sanitized delivery failure notification."""

    occurred_at: datetime
    operation: DeliveryErrorOperation
    error_type: str
    message: str
    retryable: bool
    consecutive_failures: int


@dataclass(frozen=True)
class DeliveryStatus:
    """Point-in-time delivery state plus durable queue depths."""

    enabled: bool
    worker_state: DeliveryWorkerState
    pending_events: int
    quarantined_events: int
    pending_tasks: int
    quarantined_tasks: int
    pending_outcomes: int
    quarantined_outcomes: int
    pending_revenues: int
    quarantined_revenues: int
    pending_provider_jobs: int
    quarantined_provider_jobs: int
    oldest_pending_at: datetime | None
    last_attempt_at: datetime | None
    last_success_at: datetime | None
    last_error_at: datetime | None
    last_error_type: str | None
    last_error_message: str | None
    consecutive_failures: int
    successful_batches: int
    failed_batches: int
    delivered_records: int
    backoff_seconds: float

    @property
    def pending_records(self) -> int:
        return (
            self.pending_events
            + self.pending_tasks
            + self.pending_outcomes
            + self.pending_revenues
            + self.pending_provider_jobs
        )

    @property
    def quarantined_records(self) -> int:
        return (
            self.quarantined_events
            + self.quarantined_tasks
            + self.quarantined_outcomes
            + self.quarantined_revenues
            + self.quarantined_provider_jobs
        )

    @property
    def healthy(self) -> bool:
        return (
            self.worker_state not in {"auth_failed", "backoff"}
            and self.quarantined_records == 0
        )


DeliveryErrorCallback = Callable[[DeliveryErrorEvent], None]

_CALLBACK_LOCK = threading.Lock()
_ERROR_CALLBACKS: list[DeliveryErrorCallback] = []


def on_delivery_error(callback: DeliveryErrorCallback) -> DeliveryErrorCallback:
    """Register a process-wide failure callback and return it for decorator use."""
    if not callable(callback):
        raise TypeError("delivery error callback must be callable")
    with _CALLBACK_LOCK:
        if callback not in _ERROR_CALLBACKS:
            _ERROR_CALLBACKS.append(callback)
    return callback


def remove_delivery_error_callback(callback: DeliveryErrorCallback) -> bool:
    """Remove a callback, returning whether it had been registered."""
    with _CALLBACK_LOCK:
        try:
            _ERROR_CALLBACKS.remove(callback)
        except ValueError:
            return False
    return True


def _emit_delivery_error(event: DeliveryErrorEvent) -> None:
    with _CALLBACK_LOCK:
        callbacks = tuple(_ERROR_CALLBACKS)
    for callback in callbacks:
        try:
            callback(event)
        except Exception:
            _LOG.warning("delivery error callback failed; ignoring", exc_info=True)


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _queue_counts(storage: Any) -> dict[str, Any]:
    method = getattr(storage, "delivery_counts", None)
    if not callable(method):
        return {}
    try:
        result = method()
    except Exception:
        _LOG.warning("could not read durable delivery queue counts", exc_info=True)
        return {}
    return result if isinstance(result, dict) else {}


def _count(counts: dict[str, Any], key: str) -> int:
    value = counts.get(key, 0)
    return int(value) if isinstance(value, int) and not isinstance(value, bool) else 0


def _oldest(counts: dict[str, Any]) -> datetime | None:
    value = counts.get("oldest_pending_at")
    if not isinstance(value, str):
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None or parsed.utcoffset() is None:
        return None
    return parsed.astimezone(timezone.utc)


def local_delivery_status(storage: Any | None = None) -> DeliveryStatus:
    """Build status when cloud delivery is intentionally disabled."""
    counts = {} if storage is None else _queue_counts(storage)
    return DeliveryStatus(
        enabled=False,
        worker_state="local_only",
        pending_events=_count(counts, "pending_events"),
        quarantined_events=_count(counts, "quarantined_events"),
        pending_tasks=_count(counts, "pending_tasks"),
        quarantined_tasks=_count(counts, "quarantined_tasks"),
        pending_outcomes=_count(counts, "pending_outcomes"),
        quarantined_outcomes=_count(counts, "quarantined_outcomes"),
        pending_revenues=_count(counts, "pending_revenues"),
        quarantined_revenues=_count(counts, "quarantined_revenues"),
        pending_provider_jobs=_count(counts, "pending_provider_jobs"),
        quarantined_provider_jobs=_count(counts, "quarantined_provider_jobs"),
        oldest_pending_at=_oldest(counts),
        last_attempt_at=None,
        last_success_at=None,
        last_error_at=None,
        last_error_type=None,
        last_error_message=None,
        consecutive_failures=0,
        successful_batches=0,
        failed_batches=0,
        delivered_records=0,
        backoff_seconds=0,
    )


__all__ = [
    "DeliveryErrorCallback",
    "DeliveryErrorEvent",
    "DeliveryErrorOperation",
    "DeliveryStatus",
    "DeliveryWorkerState",
    "on_delivery_error",
    "remove_delivery_error_callback",
]
